_posts/2021-05-01-monthly-dev-post.md

```diff
@@ -0,0 +1,145 @@
+We will get to this back in a bit.
+
+### Networking
+
+Mostly what stood out from the Steam release, how difficult it is to setup a network game.
+And this might sound silly, but it is not something we really realised before.
+Back in 2007, from which this network implementation dates, it was one of the only ways available.
+But as time went on, better and easier ways to join each other in a game have been invented; and we have not caught up with that.
+
+So, a few days after release, we started work on improving the whole network experience.
+This is not an easy task, as there are many moving parts.
+
+The main thing we are addressing, is how multiplayer and singleplayer are more interconnected.
+Don't worry, you still won't need an Internet connection to play OpenTTD, but we are making it more fluent.
+What this mean is, that when you start a game, you can indicate if you either want a private game (offline, nobody can join), friends-only game (your server is not listed but friends can join, see below) or a public game.
+
+- Private games are what singleplayer is now.
+- Friends-only introduces the concept of "invite codes".
+This is a short code you can give to your friends.
+They can enter it in OpenTTD to join your server.
+No fuss, nothing.
+It just does works.
+- Public games are pretty much like the current public servers.
+
+To allow this to work easier without things like port-forwarding and firewall exceptions etc etc, we will add some modern techniques to OpenTTD to make that all a bit easier.
+
+First of all, we will move away from a mix of UDP and TCP, and switch to TCP only (for now; possibly we go full UDP in the future, but baby-steps).
+Additionally, we are changing completely how the master-server (the component responsible for the server-listing) works, and rebranding it to a game-coordinator.
+
+The game-coordinator has a very important task: if player B wants to join the game of player A with the invite code, he should make that happen.
+We will use several techniques to make this happen.
+- The first is, of course, a direct connect.
+This is exactly how the current game works, and only works if you have your network setup for it.
+- The second method is a so called STUN.
+Not going into any technical details here, but it basically attempts to connect the client and the server in a smart way that doesn't require port-forwarding etc.
+The downside? It doesn't work for all network setups.
+Especially company networks are known for not allowing this method.
+- Third, and this is not 100% sure this will be added, we fall back to TURN.
+This is a method where we, as OpenTTD, relay the traffic between the client and server.
+If you have a working Internet connection, this method works.
+No setup required, no questions asked.
+The downside here is that it means we also have to pay for the traffic.
+So we have to estimate and test out how likely it is the above methods won't work; hence the "not 100% sure" part.
```

Comment:
For TURN, will there be the possibility of adding in third-party relay servers for load levelling purposes, so that the main OpenTTD relay server(s) don't need to handle as much load?

Reply:
Sorry, didn't see your comment; this is also the reason I don't like making a news topic out of this, as I want to see these questions so I can do something with them :D So tnx for asking ;)

I hope we can make this happen. In fact, I hope that if both client and server use Steam, we can always use Steam Relay over any other method. As that is just more reliable. But .. that is far away .. I don't think that happens any time soon.

So I am just making baby steps atm .. getting to a place it would be possible. Lot of work to do to get there :D

So yeah, absolutely on my wishlist.